src/pieces.rs

```diff
@@ -89,24 +54,18 @@ impl<'a> Tower<'a> {
     ///    For example, (Your) Pawn, (Your) Gold, (Your) Gold is disallowed
     ///    but (Your) Pawn, (Your) Gold, (Enemy) Gold is fine
     pub fn is_valid(&self) -> bool {
-        match (self.bottom, self.mid, self.top) {
+        use pieces::Tower::*;
+        match *self {
             // Empty towers are obviously fine!
-            (None, None, None) => true,
-            // Cases where the tower clearly isn't a tower (ie: bottom pieces missing)
-            (None, _, _) => false,
-            (Some(_), None, Some(_)) => false,
+            Empty => true,
             // Towers of just one piece can never have two pieces of the same type
-            (Some(_), None, None) => true,
-            (Some(bottom), Some(middle), None) => {
+            Single(_) => true,
             // Towers of two mustn't have the pieces be the same type and from same player
-                return !same_type_and_player(bottom, middle)
-            }
+            Double(bottom, middle) => !same_type_and_player(bottom, middle),
```

Comment:
Unless, of course, there's some other sensible meaning of `Eq`, in which case things become more complicated and you have to pick which one is the canonical one. Also, did you figure out the mutability issue for the player, since every piece is borrowing them?